[DAG siren geocodage] feat: add tests to sirene geoloc

hacherix opened the pull request:
Afin d'éviter que le DAG ne pousse des fichiers incomplets cette PR propose de rajouter des tests de cohérences sur les fichiers départements générés.

Les tests réutilisent les fichiers `stats.json` produits par le script `4b_geocode_stats.sh`.
Ils comparent les volumes et l'efficacité du géocodage avec celui du moins précédent pour s'assurer qu'aucun problème n'ai eu lieu.

J'en ai profité pour copier les fichiers qui étaient sur le serveur distant.

Dans un premier temps cette PR ne bloque pas le DAG en cas de problème remonté.
Je n'ai pas osé tester le DAG en entier.

Pour tester la nouvelle task depuis le remote:
```
cp /srv/sirene/data-sirene/2024-09/stats.json /srv/sirene/data-sirene/stats.json
/srv/sirene/venv/bin/python /srv/sirene/geocodage-sirene/4c_check_stats_coherence.py
rm /srv/sirene/data-sirene/stats.json
```

hacherix committed:
fix: add remote files in the repo for documentation

Execution files where located only on the remote machine.
There is now a copy in the repo so we can better review them.
They are sync manually so discrepancies may arise.

HAEKADI commented on data_processing/insee/sirene/geocodage/remote_files/4c_check_stats_coherence.py:
```diff
@@ -0,0 +1,149 @@
+                        f"= {delta:.2f}% (threshold: {threshold}%)"
+                    )
+
+    # Find missing fichiers
```
```suggestion
    # Find missing files
```

HAEKADI commented on data_processing/insee/sirene/geocodage/remote_files/4c_check_stats_coherence.py:
```diff
@@ -0,0 +1,149 @@
+    file_prev = load_json(file_prev)
+    file_next = load_json(file_next)
+
+    thresholds = {
+        "count": -5,  # if the new count is under -5% of the previous one then raises an error
+        "locality": -10,
+        "efficacite": -10, # efficacite = ratio of successful geocodage
+    }
```
cool!

HAEKADI reviewed (approved):
Nice!
I only reviewed the `check_stats_coherence` functioning.
The remaining geocoding logic went right over my head 😅 
If data.gouv is okay with it, we can test it this week before the end of the month and on the new data next week, before we raise an error.

Pierlou reviewed (commented):
Really nice to have the shell scripts documented here!
Once this improvement is pushed and working, maybe it will be time to clean some functions (*wink* geocode.py *wink*)

hacherix commented on data_processing/insee/sirene/geocodage/remote_files/geocode.py.py:
```diff
@@ -0,0 +1,482 @@
+#! /usr/bin/python3
```
Good catch!

hacherix commented on data_processing/insee/sirene/geocodage/remote_files/4c_check_stats_coherence.py:
```diff
@@ -0,0 +1,149 @@
+
+    missing_in_json_next = get_missing_geo_files(stats_next, stats_prev)
+    if missing_in_json_next:
+        results.append(f"[WARNING] Geo files missing in the new stats: {', '.join(missing_in_json_next)}")
```
Good question. I was not sure. I will change it to ERROR then! What is the probability that a new département would be created right?

Pierlou replied:
Or in this case erased from the map 😱

hacherix commented on data_processing/insee/sirene/geocodage/remote_files/geocode.py.py:
```diff
@@ -0,0 +1,482 @@
+import sqlite3
+import marshal
+import unidecode
+import random
```
True! I just copy pasted what's on the remote but I can fix it in both locations anyway 👌

hacherix commented on data_processing/insee/sirene/geocodage/DAG-sirene-geocodage-etalab.py:
```diff
@@ -138,6 +147,7 @@
     national_files_agregation.set_upstream(geocoding)
     prepare_to_rsync.set_upstream(national_files_agregation)
     prepare_to_rsync.set_upstream(split_by_locality)
+    check_stats_coherence.set_upstream(get_geocode_stats)
```
Good question. I am a bit afraid of making the DAG fail next month just because of a bug since I was not able to test a full execution of the DAG.
My plan was to let it run as it is on the 1st, monitor it, and if we are happy with the result to push a few changes to make it skip the rsyncs in case of failure next time.
What do you think?

Pierlou replied:
Makes sense, we go for that 🤝

hacherix committed:
feat: add tests on the geocode results

For now the tests are for information only.
Those can be changed later to trigger a DAG failure
once we have validated they behave as expected.

hacherix commented on data_processing/insee/sirene/geocodage/remote_files/4c_check_stats_coherence.py:
```diff
@@ -0,0 +1,149 @@
+                        f"[WARNING]\t{geo_file}: '{key}' is not numeric."
+                    )
+                    continue
+                if not isinstance(value_next, (int, float)):
```
True I just removed it, since we technically have control over the json output

Pierlou reviewed (approved):
Here we go 🚀
